core/parser/src/lexer/regex.rs

```diff
@@ -114,14 +117,21 @@ impl<R> Tokenizer<R> for RegexLiteral {
             }
         }
 
-        let mut flags = Vec::new();
-        let flags_start = cursor.pos();
-        cursor.take_while_ascii_pred(&mut flags, &char::is_alphabetic)?;
-
-        // SAFETY: We have already checked that the bytes are valid UTF-8.
-        let flags_str = unsafe { str::from_utf8_unchecked(flags.as_slice()) };
+        let mut flags: [Option<NonZeroU32>; MAXIMUM_REGEX_FLAGS] = [None; MAXIMUM_REGEX_FLAGS];
+        let n = cursor.take_array_alphabetic(&mut flags)?;
+        if n > MAXIMUM_REGEX_FLAGS {
+            // There can only be a maximum of 8 flags.
+            return Err(Error::syntax(
+                "Invalid regular expression: too many flags",
+                start_pos,
+            ));
+        }
+        let flags: [u32; MAXIMUM_REGEX_FLAGS] = flags.map(|c| c.map_or(0, NonZeroU32::get));
```

Comment:
I'm not sure why you bother with the `[Option<NoneZeroU32>; MAXIMUM_REGEX_FLAGS]` at all when you just convert it back to a `[u32; MAXIMUM_REGEX_FLAGS]`, it seems a bit pointless.

Why not just start with a `[u32; MAXIMUM_REGEX_FLAGS]` from the beginning, so you don't need to map it. The Option type doesn't look like it adds much in this procedure.

Reply:
As a matter of fact, I don't even think we need `u32` at all. `Option<NonZeroU8>` probably should be better.